### PR TITLE
Add support Redis lock system for commands

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) 
| Bug fix?                               | yes/no
| New feature?                           | yes/no
| Deprecations?                          | yes/no
| BC breaks?                             | yes/no
| Automated tests included?              | yes/no
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:
Mautic4 for bring support for Lock component. Now it's easy to add any of supported lock systems to our commands. This PR bring support for redis 

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Setup Redis. On ddev you can zse https://github.com/mautic/mautic/pull/10347
3. Setup it to local.php 

```php
      'cache_adapter_redis'     => [
            'dsn'     => 'redis://redis',
        ],

```
4. Run `php bin/console m:s:r -x redis`
5. Should work like before

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10347"></a>

### CODE REVIEW

File: app/bundles/CoreBundle/Command/ModeratedCommand.php [hunk 1]
```diff
@@ -180,7 +183,18 @@ private function checkPid(): bool
 
     private function checkFlock(): bool
     {
-        $store      = new FlockStore($this->runDirectory);
+        switch ($this->moderationMode) {
+            case self::MODE_REDIS:
+                $dsn   = $this->getContainer()->get('mautic.helper.core_parameters')->get('cache_adapter_redis')['dsn'] ?? null;
+                $redis = new \Redis();
+                $redis->connect(parse_url($dsn, PHP_URL_HOST));
```
Comment (mollux) [inline app/bundles/CoreBundle/Command/ModeratedCommand.php]: @kuzmany the DSN can contain a socket path, so the `parse_url` won't work in that case

Comment (kuzmany) [reply app/bundles/CoreBundle/Command/ModeratedCommand.php]: Good point 🥇 I have to check it again

Comment (kuzmany) [reply app/bundles/CoreBundle/Command/ModeratedCommand.php]: I rewrite it with support RedisAdapter , options and lock timeout

**Comment by mabumusa1** [comment]
@kuzmany we need to bring this to M5 please

**Comment by kuzmany** [comment]
@mabumusa1 I rebased it to 5.x but we added a new service CoreParametersHelper to ModerateCommand, then need the change apply to all commands and fix all unit tests related to the command.  Need more time for it.

**Comment by mabumusa1** [comment]
@kuzmany there are two failing tests, please fix them so I can test.

**Review by mabumusa1 — APPROVED** [review]
Code change looks good, I tested locally and lock worked as expected

**Review by volha-pivavarchyk — APPROVED** [review]
The command works as before

**Comment by luizeof** [comment]
Hello friends @RCheesley @escopecz , how are you?

Does this lock change modify the moderationKey ?

Mautic doesn't allow us to run a command with the --id parameter at the same time, which is great.

However, if we pass --min-contact-id or --max-contact-id, separating the contacts by groups to speed up the sending in parallel, the lock is also activated.

is there any chance to add --min-contact-id and --max-contact-id to moderationKey ?

Thank you so much for releasing 5 alpha with all your efforts to make this tool even more amazing.

**Comment by luizeof** [comment]
PS: https://github.com/mautic/mautic/pull/11549 solves a lot of things and already guarantees a lot of agility. it's more about the range of contacts.

**Comment by escopecz** [comment]
@luizeof great question! We'll be documenting this soon. Mautic 5 will send emails via [Symfony Messenger](https://symfony.com/doc/5.4/messenger.html) and so you'll be able to run as many workers to send emails concurrently as you wish. No need to split contacts into groups by ID ranges. I can't wait to read about how the community will manage auto-scaling of such workers to bring the server cost down and send emails by warp speed 🙂 🚀

**Pull request merged**